Task14.1 user badges

phantom-profile opened the pull request.

phantom-profile committed:
created db platform for badges

1) created badges table
2) connected it to users table by m2m (1 user has many achievements, 1 achievement had bu many users)
3) added seeds and default picture

phantom-profile committed:
Create img.png

phantom-profile committed:
created Badge model and associated it to User model through UserBadge model

phantom-profile committed:
added unique attr to Badge

1) there are badges which can be got once (completed all tests of category) and more (completed test with 1 attempt) so unique checks it

phantom-profile committed:
made routes for Badges

1) admin can do all CRUD with Badge
2) user can see all badges, his own and every detailed
3) object of UserBadge can be only created

phantom-profile committed:
Update routes.rb

phantom-profile committed:
made template for Controllers (without difficult logic)

phantom-profile committed:
added css for badje and little fixes

phantom-profile committed:
added views for all badge controllers methods

phantom-profile committed:
added rules to badge also made js for nice looking form

phantom-profile committed:
added comment to params

phantom-profile committed:
added links to all new resources

phantom-profile committed:
added update method

phantom-profile committed:
added check-method for each badge

phantom-profile committed:
made create action for UserBadge (when user gets badge)

VidgarVii commented on app/controllers/admin/badges_controller.rb:
```diff
@@ -0,0 +1,70 @@
+  end
+
+  def badge_params
+    valid_params = params.require(:badge).permit(:title, :level, :test_id, :category_id, :unique, :rule)
```
а если будет 100 разных правил для баджиков?
пересмотри последний созвон

VidgarVii commented on app/models/badge.rb:
```diff
@@ -0,0 +1,31 @@
+      TestPassage.where(user: user, test: rule_attribute).size == 1
+    end
+  end
+
```
получился god object
правила выдывать бадж или нет. стоит вынести в сервис

VidgarVii commented on app/models/user_badge.rb:
```diff
@@ -0,0 +1,5 @@
+class UserBadge < ApplicationRecord
+
+	belongs_to :user
+	belongs_to :badge
```
отступы съехали

VidgarVii replied:
не исправил

phantom-profile committed:
added Rule model to application

phantom-profile committed:
made own New form for rule and removed JS

phantom-profile committed:
moved rules from badge to rule

phantom-profile committed:
created controller and model for rule

phantom-profile committed:
added service for badges sending and some rules

phantom-profile committed:
removed unused parts (Rule model)

phantom-profile committed:
moved Rule functionality back to Badge

phantom-profile committed:
moved away Rule links and unused attributes

VidgarVii commented on app/models/badge.rb:
```diff
@@ -0,0 +1,13 @@
+    'Completed tests of category': BadgeRules::AllTestsOfCategory,
+    'Completed tests of level': BadgeRules::AllTestsOfLevel,
+    'Completed test with first attempt': BadgeRules::TestWithFirstAttempt
+  }.freeze
```
а надо ли модельке знать про сами классы спецификаций? Спецификации вызывает тот кто выдаёт баджи

VidgarVii commented on app/models/badge.rb:
```diff
@@ -0,0 +1,13 @@
+
+  has_many :user_badges, dependent: :destroy
+  has_many :users, through: :user_badges
+end
```
не хватает валидаций

VidgarVii replied:
rule_attribute не всегда нужен

VidgarVii commented on app/services/badge_rules/all_tests_of_category.rb:
```diff
@@ -0,0 +1,8 @@
+    category = @badge.rule_attribute
+    tests_of_category = Test.where(category: category)
+    user.tests & tests_of_category == tests_of_category
+  end
```
метод по факту вроде как должне быть предикатным. А вот по названию и не скажешь. call принято использовать для получения какого-то объекта с данными.

VidgarVii commented on app/services/badge_rules/all_tests_of_category.rb:
```diff
@@ -0,0 +1,8 @@
+    user = @test_passage.user
+    category = @badge.rule_attribute
+    tests_of_category = Test.where(category: category)
+    user.tests & tests_of_category == tests_of_category
```
&&

VidgarVii replied:
а нам точно надо всегда проверять у юзера наличие тестов. И это точно пройденные тесты? по названию не скажешь.

VidgarVii commented on app/services/badge_rules/base_badge_rule.rb:
```diff
@@ -0,0 +1,15 @@
+    return false if badge.unique && badge.users.include?(test_passage.user)
+
+    new(test_passage, badge).call
+  end
```
а тут на оборот. achieved? явно предикат. А вот call не понятно

VidgarVii commented on app/services/badge_sending.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  def call
+    puts(Badge::RULES.inspect)
```
путсы надо удалять. Вооще лучше и точней использовать дебагер

VidgarVii commented on app/services/badge_sending.rb:
```diff
@@ -0,0 +1,22 @@
+    puts(Badge::RULES.inspect)
+    Badge.find_each do |badge|
+      if Badge::RULES[badge.rule.to_sym].achieved?(@test_passage, badge)
+        UserBadge.create(badge: badge, user: @user)
```
а как же методы ассоциаций?
без них не со всем понятно что тут вообще происходит.
по моему на созвоне я классный метод придложил. add_badge(badge)

VidgarVii commented on app/services/badge_sending.rb:
```diff
@@ -0,0 +1,22 @@
+    Badge.find_each do |badge|
+      if Badge::RULES[badge.rule.to_sym].achieved?(@test_passage, badge)
+        UserBadge.create(badge: badge, user: @user)
+        @message = "#{badge.title} achieved"
```
@message где-то разве используется?

VidgarVii replied:
нашел. установка сообщений во flash
это дикий side effect
метод должен делать что-то одно.

phantom-profile committed:
2 spaces instead of tabs

phantom-profile committed:
replaced hash with array and hash moved to service

also removed puts and used associations
and small fix in controller due to array using
added two useful scopes

phantom-profile committed:
renamed methods in rule from call to "?"

because they return true/false but not the objects
also used scopes

phantom-profile committed:
useless scope
